From fd2702c494f13802bb56efbcc17d9ca3ca9d61e7 Mon Sep 17 00:00:00 2001
From: Mitul Chokshi <mitul.chokshi@intel.com>
Date: Thu, 4 Jan 2018 21:12:34 +0800
Subject: [PATCH 0490/1240] drm/i915/gvt: Add vgt-id in context id

Context id is reported to OABUFFER along with performance statistics
counter. So when performance monitoring code gets the Context id, it
can extract vm-id to identify which VM submitted given context.
v2: define SIZE_CONTEXT_HW_ID even when CONFIG_DRM_I915_GVT is not set,
which fixes RTC defect 192523.

Change-Id: I7e8ed7f741e2f41e2e9da4b7bdd463dfd3e2fe12
Signed-off-by: Mitul Chokshi <mitul.chokshi@intel.com>
Signed-off-by: Daniel van der Wath <danielx.j.van.der.wath@intel.com>
Reviewed-by: Singh, Satyeshwar <satyeshwar.singh@intel.com>
Reviewed-by: Adebisi, YetundeX <yetundex.adebisi@intel.com>
Verified-by: Van Der Wath, Daniel J <danielx.j.van.der.wath@intel.com>
(cherry picked from commit 3e52d8f679eb93c2c1eb7e0f69de34c7d78260d2)
(cherry picked from commit 74c64f2595b70a2079c28bbda960fff7d3e99289)
Reviewed-by: Jiang, Fei <fei.jiang@intel.com>
Reviewed-by: He, Min <min.he@intel.com>
Reviewed-by: Dong, Eddie <eddie.dong@intel.com>
Tested-by: Dong, Eddie <eddie.dong@intel.com>
---
 drivers/gpu/drm/i915/i915_drv.h         | 8 +++++++-
 drivers/gpu/drm/i915/i915_gem_context.c | 6 +++++-
 2 files changed, 12 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index 92c9751..ba3a171 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -2396,7 +2396,13 @@ struct drm_i915_private {
 		 * This is limited in execlists to 21 bits.
 		 */
 		struct ida hw_ida;
-#define MAX_CONTEXT_HW_ID (1<<21) /* exclusive */
+#ifdef CONFIG_DRM_I915_GVT
+	/* In case of virtualization, 3-bits of vgt-id will be added to hw_id */
+#define SIZE_CONTEXT_HW_ID (18)
+#else
+#define SIZE_CONTEXT_HW_ID (21)
+#endif
+#define MAX_CONTEXT_HW_ID (1<<SIZE_CONTEXT_HW_ID) /* exclusive */
 	} contexts;
 
 	u32 fdi_rx_config;
diff --git a/drivers/gpu/drm/i915/i915_gem_context.c b/drivers/gpu/drm/i915/i915_gem_context.c
index 8afd2ce..8a6dec9 100644
--- a/drivers/gpu/drm/i915/i915_gem_context.c
+++ b/drivers/gpu/drm/i915/i915_gem_context.c
@@ -146,7 +146,11 @@ static void i915_gem_context_free(struct i915_gem_context *ctx)
 
 	list_del(&ctx->link);
 
-	ida_simple_remove(&ctx->i915->contexts.hw_ida, ctx->hw_id);
+	if (intel_vgpu_active(ctx->i915))
+		ida_simple_remove(&ctx->i915->contexts.hw_ida, ctx->hw_id & ~(0x7 << SIZE_CONTEXT_HW_ID));
+	else
+		ida_simple_remove(&ctx->i915->contexts.hw_ida, ctx->hw_id);
+
 	kfree_rcu(ctx, rcu);
 }
 
-- 
2.7.4

